src/System.Runtime.Extensions/src/System/Runtime/CompilerServices/SwitchExpressionException.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates that a switch expression that was non-exhaustive failed to match its input
+    /// at runtime, e.g. in the C# 8 expression <code>3 switch { 4 => 5 }</code>.
+    /// The exception optionally contains an object representing the unmatched value.
+    /// </summary>
+    [Serializable]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    public sealed class SwitchExpressionException : InvalidOperationException
+    {
+        public SwitchExpressionException()
+            : base(SR.Arg_SwitchExpressionException) { }
+
+        public SwitchExpressionException(object unmatchedValue)
+            : this()
+        {
+            UnmatchedValue = unmatchedValue;
+        }
+
+        private SwitchExpressionException(SerializationInfo info, StreamingContext context)
+            : base(info, context)
+        {
+            if (info == null)
+            {
+                throw new ArgumentNullException(nameof(info));
+            }
+            UnmatchedValue = info.GetValue(nameof(UnmatchedValue), typeof(object));
+        }
+
+        public object UnmatchedValue { get; }
+
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            base.GetObjectData(info, context);
+            if (info == null)
+            {
+                throw new ArgumentNullException(nameof(info));
+            }
+            info.AddValue(nameof(UnmatchedValue), UnmatchedValue, typeof(object));
+        }
+
+        public override string Message
+        {
+            get
+            {
+                string s = base.Message;
+                if (UnmatchedValue != null)
```

Comment:
Is there some guidelines for new code and use of `is` or `!(val is null)` @danmosemsft @stephentoub

Reply:
@stephentoub @danmosemsft @ahsonkhan 
the "is null" check is it too much unnecessary work than just a direct compare?

Reply:
It's identical, so keeping `is null` is fine.

[sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMA3ALABQWAzAQEx4DCeA3hXuwTQJYB2MeAWUIAKIvigBKFmw6yuAMzzCoeLip4BXADZaJM2e1bkDJggHY8AWhL6DAX1uysFqADoAMnB4BzGAAsyYw4HIPZbalU+QVpRQnEpI1NVRWU8AF40vE0dPVCDRKSOZysbPNkQpOK3Tx9/QPKKOyA==)

Reply:
AFAIK it depends on "s" type [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACgcBmIgJgIGECBvKgzoutAOxgIBZYgAoAKgSgBKNlzkEO8zmgBmBEVAJpNvAK4YMUxUvaUlSnAHYCAWjLH5AXwdyrkgHQAZOLwDmMAAsKMy5nEM4HWi1+IXpxSRlWcwVw+VV1TQBeTII9AyNUuVNk12s7YOSw5LcoLx9/IIcwsKoonEYxKmKuKL4Bbz9Atl84GFIoUdIqrki6EgA2AmAAe2WMAmWAB0QIGGWEbPUJZeAAK2IkAmOz+hlzB26LaxgEXTgK0NmiYkWVtY3tghdvsAIQ5eInc6Xa6nW4PFw9Z6vd5NKiOIA===) if type overload `==` `!=` operators compiler emit different IL